### ISSUE TITLE
nixos 18.09: upgrade xorg to 1.20

### ISSUE THREAD

**Issue by volth** [NixOS/nixpkgs#44492]
## Issue description

xorg 1.20 is 3 month old and [many distros](https://repology.org/metapackage/xorg-server/versions) have already switched to it.

**Comment by etu** [comment]
Sidenote: 1.20.1 is out with bug fixes: https://lists.freedesktop.org/archives/xorg/2018-August/059380.html

**Comment by volth** [comment]
I tried to use [the update script ](https://github.com/NixOS/nixpkgs/blob/b7ba5dec0d8027980012227e32855d71f0fa888d/pkgs/servers/x11/xorg/generate-expr-from-tarballs.pl#L3) but it downgraded many tarballs.

**Comment by vcunat** [comment]
You don't need to do that line.  I always just edited the versions manually, and the script does the rest.

**Comment by NeQuissimus** [comment]
I looked at this for a while but I think I am in over my head.
I made a few changes but the derivation generation reports **a lot** of `NOT FOUND` that I cannot figure out how to resolve:

https://gist.github.com/NeQuissimus/9e11fc78f5c986d11e21a5c1cca80a37

**Comment by dezgeg** [comment]
I wonder if there is much need for the update script anymore as long as the derivations would be written in a way that @r-ryantm can find updates for them?

**Comment by ryantm** [comment]
nixpkgs-update/r-ryantm currently skips xorg because the expressions [have multiple fetchers](https://gist.github.com/ryantm/ac41cddb2857da6f55abab6ae5037e4c#file-20180817_ups-log-L12) in the same file. If each one was written as a separate file it would start trying to update them.

I'm not sure that it is needed, but it doesn't have a way of grouping updates.

**Comment by vcunat** [comment]
The `NOT FOUND` errors are normal for the script IIRC, and the resulting diff seems OK at a superficial look.

**Comment by xeji** [comment]
With the update, it would be great to configure X to be able to run as non-root (without a display manager, started by a normal user with `startx`). Not exactly sure what needs to be changed for that to work.

**Comment by xeji** [comment]
Not necessarily. I don't mean "always run as non-root" but "be able to run as non-root". E.g. on Arch, you can either use a display manager which runs X as root, or start it as a normal user, both works.

**Comment by xaverdh** [comment]
@xeji
There are instructions along that line in https://nixos.wiki/wiki/Using_X_without_a_Display_Manager.

Those are not very nixy though, so I wrote a small module (https://gitlab.com/xaverdh/my-nixos-config/blob/master/xserver-rootless.nix) which does just enough, to make things work with my setup.
Its probably horribly broken in more than one way, but it "works for me".

**Comment by jb55** [comment]
so I'm running xorg 1.20.1 on 4.18.11 now. Things appear to work. I bypassed the update script for now as it probably needs to be rewritten/modified. My reasoning is that their website says R7.7 was the last official release and from then on things should be updated on an as-is basis. So I decided to just update the proto stuff and bump xorgserver to 1.20.1.

Will start putting a patch together.

**Comment by jb55** [comment]
My main motivation for this was the new drm lease support, used by SteamVR for high performance VR in linux. This requires 4.18 so I figure I would bump that as well. I'm testing right now to make sure that works. It looks like a need to update and test a few more things before I put a patch together.

**Comment by vcunat** [comment]
4.18 doesn't seem suitable as kernel default.  The policy so far is to use the latest longterm kernel (once it's in some fairly stable state), and that will soon be 4.19.  I'd expect we can switch the default to 4.19 on unstable in several weeks.

**Comment by jb55** [comment]
Vladimír Čunát <notifications@github.com> writes:
> 4.18 doesn't seem suitable as kernel default.  The policy so far is to use the latest longterm kernel (once it's in some fairly stable state), and that will soon be 4.19.  I'd expect we can switch the default to 4.19 on unstable in several weeks.

Good to know!

I'm working on this branch for anyone interesting in testing this out:

    git fetch https://github.com/jb55/nixpkgs xorg-1.20
    git checkout -b xorg-1.20 FETCH_HEAD

**Comment by vcunat** [comment]
@jb55 `pkgs/servers/x11/xorg/default.nix` is still (almost whole) a generated file.  Your changes to it appear "manual", so that will clash with future updates.  Well we might change the style we do this, but so far the generation seemed to work fine to me.  What are your plans about that?

**Comment by xeji** [comment]
@xaverdh thank you, I'll give that a try! We definitely should work on a NixOS module that allows using xorg without a display manager.

**Comment by jb55** [comment]
Vladimír Čunát <notifications@github.com> writes:

> @jb55 `pkgs/servers/x11/xorg/default.nix` is still (almost whole) a generated file.  Your changes to it appear "manual", so that will clash with future updates.  Well we might change the style we do this, but so far the generation seemed to work fine to me.  What are your plans about that?

It's a generated file, but the releases it was generating from are no
longer updated. The site mentions that there may be a 7.8 release in the
future, but noone has stepped up to be a release manager:

https://www.x.org/wiki/Releases/7.8/

I think it makes sense to update them individually for now until a new
release is available? The script would probably need to be updated at
that time as well.

**Comment by vcunat** [comment]
Well, yes, you need to update the version you want manually, but the rest still works.

**Comment by jb55** [comment]
@vcunat ok I think I see what you mean. In theory if I just scrape all the latest tarballs it should work?

**Comment by vcunat** [comment]
Just their names, that's enough.  It was done recently but not finished, see https://github.com/NixOS/nixpkgs/issues/44492#issuecomment-414157422 (above)

**Comment by vcunat** [comment]
Thinking longterm, I should explicitly mention another viable alternative: migrate to the usual one file per package (without override.nix) and use some "universal updater" instead, e.g. r-ryantm.

**Comment by vcunat** [comment]
> Really would be nice to be able to use a modern X server! :grin:

Wayland compositor, you mean?

**Comment by jb55** [comment]
Will Dietz <notifications@github.com> writes:

> The discussion about auto-generation is a bit orthogonal to your
> changes here, I think: the big pain here was the unification of the
> proto's into xorgproto which is a big manual change and not something
> any update system (new or existing) would handle anyway.

agreed, but @vcunat claims that it should still be possible, I just
haven't had the time to look into updating the script.

> Can you submit this a PR? :+1: Or is it unsuitable for some reason?

mostly blocked on the above point.

-- 
https://jb55.com

**Comment by volth** [comment]
There a security vulnerability in Xorg prior to ```1.20.3```
https://www.opennet.ru/opennews/art.shtml?num=49500
